sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -847,7 +847,117 @@ protected BigtableSource withEstimatedSizeBytes(Long estimatedSizeBytes) {
           Math.max(sizeEstimate / maximumNumberOfSplits, desiredBundleSizeBytes);
 
       // Delegate to testable helper.
-      return splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      List<BigtableSource> splits =
+          splitBasedOnSamples(desiredBundleSizeBytes, getSampleRowKeys(options));
+      return reduceSplits(splits, options);
+    }
+
+    private List<BigtableSource>
+        reduceSplits(List<BigtableSource> splits, PipelineOptions options) throws IOException {
+      final long maximumCountOfSplits = 15_360L;
+      int numberToCombine =
+          (int) ((splits.size() + maximumCountOfSplits - 1) / maximumCountOfSplits);
+      if (splits.size() < maximumCountOfSplits || numberToCombine < 2) {
+        return splits;
+      }
+      ImmutableList.Builder<BigtableSource> reducedSplits = ImmutableList.builder();
+      ImmutableList.Builder<ByteKeyRange> mergedRanges = ImmutableList.builder();
+      int counter = 0;
+      long size = 0;
+      List<ByteKeyRange> previousSourceRanges = null;
+      for (BigtableSource source : splits) {
+        counter++;
+        if (checkRangeAdjacency(previousSourceRanges, source.getRanges())) {
+          mergedRanges.addAll(source.getRanges());
```

Comment:
Can you please collapse the ranges here?  If everything's adjacent, than there's no need to keep on checking the same ranges in the future.

Reply:
Yep will use mergeRanges to collapse previousRanges.